src/PayerRegistry.sol

```diff
@@ -0,0 +1,530 @@
+        address feeDistributor_,
+        uint96 minimumDeposit_,
+        uint32 withdrawLockPeriod_,
+        int104 minActiveBalance_
+    ) external initializer {
+        require(admin_ != address(0), ZeroAdminAddress());
+
+        __UUPSUpgradeable_init();
+        __Pausable_init();
+
+        _setAdmin(admin_);
+        _setSettler(settler_);
+        _setFeeDistributor(feeDistributor_);
+        _setMinimumDeposit(minimumDeposit_);
+        _setWithdrawLockPeriod(withdrawLockPeriod_);
+        _setMinActiveBalance(minActiveBalance_);
+    }
+
+    /* ============ Interactive Functions ============ */
+
+    /// @inheritdoc IPayerRegistry
+    function deposit(address payer_, uint96 amount_) external whenNotPaused {
+        _deposit(msg.sender, payer_, amount_);
+    }
+
+    /// @inheritdoc IPayerRegistry
+    function deposit(uint96 amount_) external whenNotPaused {
+        _deposit(msg.sender, msg.sender, amount_);
+    }
+
+    /// @inheritdoc IPayerRegistry
+    function requestWithdrawal(uint96 amount_) external whenNotPaused {
+        require(amount_ > 0, ZeroWithdrawalAmount());
+
+        // slither-disable-next-line unused-return
+        _cancelWithdrawal(msg.sender); // Cancel withdrawal (returning pending withdrawal to balance) if needed.
+
+        PayerRegistryStorage storage $ = _getPayerRegistryStorage();
+        Payer storage payer_ = $.payers[msg.sender];
+
+        payer_.pendingWithdrawal = amount_;
+        payer_.withdrawableTimestamp = uint32(block.timestamp) + $.withdrawLockPeriod;
+
+        emit WithdrawalRequested(msg.sender, amount_, payer_.withdrawableTimestamp);
+
+        uint96 debtIncurred_ = _decreaseBalance(msg.sender, amount_, minActiveBalance());
+
+        require(debtIncurred_ == 0, InsufficientBalance());
+    }
+
+    /// @inheritdoc IPayerRegistry
+    function cancelWithdrawal() external whenNotPaused {
+        require(_cancelWithdrawal(msg.sender), NoPendingWithdrawal()); // Revert if no pending withdrawal to cancel.
+    }
+
+    /// @inheritdoc IPayerRegistry
+    function finalizeWithdrawal(address recipient_) external whenNotPaused {
+        PayerRegistryStorage storage $ = _getPayerRegistryStorage();
+        Payer storage payer_ = $.payers[msg.sender];
+        uint96 pendingWithdrawal_ = payer_.pendingWithdrawal;
+
+        require(pendingWithdrawal_ > 0, NoPendingWithdrawal());
+        require(payer_.balance >= 0, PayerInDebt());
+
+        // slither-disable-next-line timestamp
+        require(
+            block.timestamp >= payer_.withdrawableTimestamp,
+            WithdrawalNotReady(uint32(block.timestamp), payer_.withdrawableTimestamp)
+        );
+
+        delete payer_.pendingWithdrawal;
+        delete payer_.withdrawableTimestamp;
+
+        $.totalDeposits -= _toInt104(pendingWithdrawal_);
+
+        emit WithdrawalFinalized(msg.sender);
+
+        require(ERC20Helper.transfer(token, recipient_, pendingWithdrawal_), ERC20TransferFailed());
+    }
```

Comment:
It can happen that the payer wants to withdraw an amount while still publishing messages to the network, and eventually going into debt.
If that's the case, we should settle any debts before finalizing the withdrawal.

Reply:
The withdrawal is blocked if they are in debt via `require(payer_.balance >= 0, PayerInDebt());`. Keeping in line with `Avoid Salvaging Transactions or Intents`, the withdrawal should finalize as expected (given that their balance was already decremented from in `requestWithdrawal` ) or fail, forcing them to cancel it or deposit more funds.

Reply:
Good, missed that require.